### PR TITLE
Add tests for new header expose

### LINKED ISSUE #3: [Controller] Allow to access request headers
Expose a `request` object with the `headers` property to allow users to access request headers. Users should be able to access headers data by both meta-variable names (e.g. `HTTP_CONTENT_TYPE`) and original names (e.g. `Content-Type`).

Example:

```ruby
request.headers["Content-Type"]
```

Make sure to thoroughly cover the change with tests and add documentation using [YARD](https://rubydoc.info/gems/yard/file/docs/Tags.md). Also, please don't use the `Rack::Request` object.

💡 The change should be made in [rage/controller/api.rb](https://github.com/rage-rb/rage/blob/master/lib/rage/controller/api.rb).
💡 Use `yard server --reload` to verify your documentation changes.

❗ Do not hesitate to use [benchmark-ips](https://github.com/evanphx/benchmark-ips) whenever there are different implementation options. When working on gems, performance is much more important than readability.

***Please comment on this issue or assign to yourself in case you have started to work on it.***

### PR DESCRIPTION
**Pull request by arikarim**
Resolves #3

### CODE REVIEW

**Comment by arikarim** [comment]
@rsamoilov sorry for the bothering :)

File: lib/rage/controller/api.rb [hunk 1]
```diff
@@ -283,4 +285,33 @@ def headers
     @__headers = {}.merge!(@__headers) if DEFAULT_HEADERS.equal?(@__headers)
     @__headers
   end
+
+  class Request
```
Comment (rsamoilov) [inline lib/rage/controller/api.rb:289]: Unfortunately, it's a bit more complicated.

We can't iterate through the whole env normalizing header names, as it is inefficient. A user might only need one header value, but your implementation will loop through all of them to do that.

On top of that, the `Content-Type` and `Content-Length` headers are not getting the `HTTP_` prefix.

Instead, nothing should be done until an actual header is requested. When it is, the code should look something like this:

```ruby
if requested_header.start_with?("HTTP_")
  env[requested_header]
else
  normalized_name = ... # normalize the header name, e.g. User-Agent -> HTTP_USER_AGENT
  env[normalized_name]
end
```

Also, the code should somehow make exceptions for the two Content headers.

**Comment by arikarim** [comment]
Hey @rsamoilov  I did what I could, please feel free to close the pull request if you felt that you need another approach.

File: lib/rage/controller/api.rb [hunk 2]
```diff
@@ -283,4 +288,42 @@ def headers
...
+    end
+
+    class Headers
+      def initialize(env)
+        @env = env
+        @headers = {}
+      end
+
+      def [](requested_header)
+        requested_header = requested_header.upcase.tr('-', '_')
```
Comment (rsamoilov) [inline lib/rage/controller/api.rb:315]: so I've run this benchmark:

```ruby
require "benchmark/ips"

requested_header = "Content-Type".freeze

Benchmark.ips do |x|
  x.report("1") { requested_header.upcase.tr!("-", "_") }
  x.report("2") { s = requested_header.tr("-", "_"); s.upcase! || s }

  x.compare!
end && true
```

and received the following result:

```
Warming up --------------------------------------
                   1   308.241k i/100ms
                   2   442.991k i/100ms
Calculating -------------------------------------
                   1      3.043M (± 1.5%) i/s -     15.412M in   5.066559s
                   2      4.575M (± 0.5%) i/s -     23.036M in   5.035321s

Comparison:
                   2:  4574912.3 i/s
                   1:  3042613.1 i/s - 1.50x  slower
```

Based on this, let's move this line inside the `else` branch and change it accordingly.

File: lib/rage/controller/api.rb [hunk 3]
```diff
@@ -283,4 +288,42 @@ def headers
...
+
+        if requested_header.start_with?("HTTP_")
+          @env[requested_header]
+        else
+          normalized_name = "HTTP_" + requested_header
+          if SPECIAL_HEADERS.include?(requested_header)
+            normalized_name = requested_header
+          end
+          @env[normalized_name]
```
Comment (rsamoilov) [inline lib/rage/controller/api.rb:324]: again, for performance reasons, let's update this:

```suggestion
          if "CONTENT_TYPE" == requested_header || "CONTENT_LENGTH" == requested_header
            @env[requested_header]
          else
            (requested_header = requested_header.tr('-', '_')).upcase!
            @env["HTTP_#{requested_header}"]
          end
```

File: lib/rage/controller/api.rb [hunk 4]
```diff
@@ -283,4 +288,42 @@ def headers
...
+    end
+
+    def headers
+      @headers ||= Headers.new(@env)
+    end
+
+    class Headers
+      def initialize(env)
+        @env = env
+        @headers = {}
```
Comment (rsamoilov) [inline lib/rage/controller/api.rb:311]: This variable seems to be not needed.

File: lib/rage/controller/api.rb [hunk 5]
```diff
@@ -283,4 +288,42 @@ def headers
     @__headers = {}.merge!(@__headers) if DEFAULT_HEADERS.equal?(@__headers)
     @__headers
   end
+
+  class Request
+    SPECIAL_HEADERS = %w[CONTENT_TYPE CONTENT_LENGTH].freeze
```
Comment (rsamoilov) [inline lib/rage/controller/api.rb:293]: Not needed.

File: lib/rage/controller/api.rb [hunk 6]
```diff
@@ -214,13 +214,18 @@ def skip_before_action(action_name, only: nil, except: nil)
   DEFAULT_HEADERS = { "content-type" => "application/json; charset=utf-8" }.freeze
 
   # @private
+  attr_reader :request
```
Comment (rsamoilov) [inline lib/rage/controller/api.rb:217]: Not needed.

**Review by rsamoilov — COMMENTED** [review]
This is exactly what I've had in my mind! Thank you @arikarim .

I've left some comments, but it's already almost perfect.

**Comment by arikarim** [comment]
Thanku for the comments, It should be good now

**Review by rsamoilov — APPROVED** [review]
Merged! Thank you!

**Pull request merged**